lib/stream_topic.js

```diff
@@ -0,0 +1,158 @@
+var minimatch = require('minimatch');
+
+var StreamTopic = module.exports = function() {
+  this.serverName = null;
+  this.deviceType = null;
+  this.deviceId = null;
+  this.streamName = null;
+  this.streamQuery =  null;
+  this._original = null;
+  this._useComponents = false;
+}
+
+StreamTopic.prototype.parse = function(topicString){
+  this._original = topicString;
+
+  var previousCharacter = null;
+  var currentCharacter = null;
+  var start = 0;
+  var topicComponents = [];
+  for(var i = 0; i < topicString.length; i++) {
+    currentCharacter = topicString[i];
+    if(currentCharacter === '/' && previousCharacter !== '\\') {
+      topicComponents.push(topicString.slice(start, i));
+      start = i + 1;
+    } else if(i === topicString.length - 1) {
+      topicComponents.push(topicString.slice(start, topicString.length));
+    }
+    previousCharacter = currentCharacter;
+  }
+
+  if (topicComponents.length < 3 && topicComponents.indexOf('**') === -1) {
+    return;
+  }
+
+  if (topicComponents.length === 3) {
+    topicComponents.unshift(null);
+  }
+
+  function checkForRegex(s) {
```

Comment:
Should we auto-insert a beginning `^` and a trailing `$`?  I think this is what argo-url-router does for regular expressions in routes.  It's also what happens in smaller subsets of Regular Expression syntax, such as what's found in XML Schema.